### ISSUE TITLE
Call to dpgettext breaks contract

### ISSUE THREAD

**Issue by alessandrousseglioviretta** [elixir-gettext/gettext#233]
Dialyxir complains that the call to `dpgettext`in the `dgettext` function in `lib/gettext.ex`

```
def dgettext(backend, domain, msgid, bindings \\ %{}) do
    dpgettext(backend, domain, nil, msgid, bindings)
end
```

breaks the following contract:

`@spec dpgettext(module, binary, binary, binary, bindings) :: binary`

Being  `nil`not `binary`, it seems that the contract is broken indeed.

**Comment by whatyouhide** [comment]
@alessandrousseglioviretta could you link to the code that breaks the contract? I'll be on mobile only for a couple days 😁, so it would help!

**Comment by josevalim** [comment]
A pull request would be even better! Generally speaking we don’t use Dialyzer, so unless someone provides a fix, it is unlikely we will tackle it ourselves. :)

**Comment by alessandrousseglioviretta** [comment]
I fixed the issue with the call to `dpgettext`, but there are still a few issues, like  `Extractor.create_translation_struct` that returns a _struct that differs from the success typing arguments_, which do not seem trivial to fix. I think it's worthwhile to fix them, but I'll need some support.

**Comment by whatyouhide** [comment]
@alessandrousseglioviretta could you provide more insight on your fix and on the issues you mentioned?

**Comment by alessandrousseglioviretta** [comment]
changed the spec for `dpgettext` from:

`@spec dpgettext(module, binary, binary, binary, bindings) :: binary`

to:

`@spec dpgettext(module, binary, binary | nil, binary, bindings) :: binary`

and , in type definition for `Gettext.PO.Translation` and `Gettext.PO.PluralTranslation`, turned

`po_source_line: pos_integer`

to

`po_source_line: pos_integer | nil`

**Comment by josevalim** [comment]
Can you please send a PR with those fixes? Thank you!

**Comment by alessandrousseglioviretta** [comment]
I managed to make Dialyzer happy (including issue #234) by modifying the default values for `Gettext.PO.Translation` and `Gettext.PO.PluralTranslation`. It's all fine and dandy when I use my fork in my app, but I could not ascertain whether the default values I chose make sense. I made also a few minor enhancements (always under the guidance of Dialyzer) and I removed a incorrect spec - in a very unorthodox way - without replacing it by the spec suggested by Dialyxir. Some 20 tests are now broken, most of them because of some extra newlines, which is not very worrying. Before fixing the tests, I'd like to have Jose's and Andre's blessing regarding my modifications. PR arriving soon.
